modules/gdscript/gdscript_parser.cpp

```diff
@@ -4171,6 +4173,48 @@ bool GDScriptParser::onready_annotation(AnnotationNode *p_annotation, Node *p_ta
 	return true;
 }
 
+bool GDScriptParser::tool_button_annotation(AnnotationNode *p_annotation, Node *p_target, ClassNode *p_class) {
+#ifdef TOOLS_ENABLED
+	ERR_FAIL_COND_V_MSG(p_target->type != Node::FUNCTION, false, R"("@tool_button" annotation can only be applied to functions.)");
+
+	FunctionNode *func = static_cast<FunctionNode *>(p_target);
+
+	if (!is_tool()) {
+		push_error("Tool buttons can only be used in tool scripts (add `@tool` to the top of the script).", p_annotation);
+		return false;
+	}
+
+	if (func->has_tool_button) {
+		push_error(vformat(R"("%s" annotation can only be used once per function.)", p_annotation->name), p_annotation);
+		return false;
+	}
+
+	p_annotation->export_info.name = "@tool_button_" + func->identifier->name;
+	p_annotation->export_info.type = Variant::Type::CALLABLE;
+	p_annotation->export_info.usage = PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_INTERNAL;
+
+	if (func->parameters.size() >= 1 && func->parameters[0]->datatype_specifier != nullptr && func->parameters[0]->datatype_specifier->type_chain[0]->name != "EditorUndoRedoManager") {
+		push_error("Tool button methods should have their first argument typed as an EditorUndoRedoManager.", func);
+		return false;
+	}
+
+	// Build the hint string (format: "<method>","<text>","[icon]")
+	String hint_string = func->identifier->name;
+	if (p_annotation->resolved_arguments.size() > 0) {
+		hint_string += "," + p_annotation->resolved_arguments[0].operator String(); // Button text.
+	}
+	if (p_annotation->resolved_arguments.size() > 1) {
+		hint_string += "," + p_annotation->resolved_arguments[1].operator String(); // Button icon.
+	}
+	p_annotation->export_info.hint_string = hint_string;
+
+	current_class->add_tool_button_member(p_annotation);
```

Comment:
This violates our assumption that class member can only be added at parser stage (annotations are resolved and applied in the analyzer). I caught the "Reaching unhandled type." error because of this. When `Vector` resizes, the container can be moved to another memory location, and in the analyzer we have a home-made `Finally` template with a lambda that holds a reference to the old memory. CC @rune-scape

https://github.com/godotengine/godot/blob/f7c567e2f56d6e63f4749387a67e5ea4903c4696/modules/gdscript/gdscript_analyzer.cpp#L925-L937

```
#0  GDScriptParser::ClassNode::Member::get_datatype (this=0x55557ed5cf50) at modules/gdscript/gdscript_parser.h:677
#1  0x00005555591acb27 in operator() (__closure=0x7fffffff94d0) at modules/gdscript/gdscript_analyzer.cpp:932
#2  0x00005555591d0008 in GDScriptAnalyzer::Finally<GDScriptAnalyzer::resolve_class_member(GDScriptParser::ClassNode*, int, const GDScriptParser::Node*)::<lambda()> >::~Finally(void) (this=0x7fffffff94d0, __in_chrg=<optimized out>) at modules/gdscript/gdscript_analyzer.h:52
#3  0x00005555591aeca5 in GDScriptAnalyzer::resolve_class_member (this=0x7fffffff9c70, p_class=0x555562468560, p_index=0, p_source=0x55557ed56c60) at modules/gdscript/gdscript_analyzer.cpp:1216
#4  0x00005555591af12a in GDScriptAnalyzer::resolve_class_interface (this=0x7fffffff9c70, p_class=0x555562468560, p_source=0x555562468560) at modules/gdscript/gdscript_analyzer.cpp:1268
#5  0x00005555591af397 in GDScriptAnalyzer::resolve_class_interface (this=0x7fffffff9c70, p_class=0x555562468560, p_recursive=true) at modules/gdscript/gdscript_analyzer.cpp:1296
#6  0x00005555591cfd1b in GDScriptAnalyzer::resolve_interface (this=0x7fffffff9c70) at modules/gdscript/gdscript_analyzer.cpp:6216
#7  0x00005555591cfee9 in GDScriptAnalyzer::analyze (this=0x7fffffff9c70) at modules/gdscript/gdscript_analyzer.cpp:6250
```

I think that instead of applying `@tool_button` to the function, we could make it standalone, like `@export_{category,group,subgroup}`. It would just require a mandatory `method` argument.

```gdscript
@tool_button(method: StringName, text: String = "", icon: StringName = "")
```

The upside is that it would give you the ability to change the button position in the inspector (currently buttons are always shown at the end, even if you move the function up). Imagine you have a lot of properties and want to hide some buttons inside (sub)groups. I wanted to suggest this even before I found the bug.

```gdscript
@tool_button("my_method_1", "Button Before")
@export var a: int
@tool_button("my_method_2", "Button After")
```

Reply:
Yikes, hidden assumptions like this might need a dev mode assert. Well, I've went with your suggestions to change to `@tool_button` so this should be ok now!